pkg/controllers/user/pipeline/remote/github/github.go

```diff
@@ -287,6 +289,27 @@ func (c *client) GetDefaultBranch(repoURL string, accessToken string) (string, e
 
 }
 
+func (c *client) GetHeadCommit(repoURL string, branch string, credential *v3.SourceCodeCredential) (string, error) {
+
+	if credential != nil {
+		userName := credential.Spec.LoginName
+		token := credential.Spec.AccessToken
+		repoURL = strings.Replace(repoURL, "://", "://"+userName+":"+token+"@", 1)
+	}
+
+	return getBranchHeadCommit(repoURL, branch)
+}
+
+func getBranchHeadCommit(url, branch string) (string, error) {
+	cmd := exec.Command("git", "ls-remote", url, branch)
```

Comment:
@gitlawr is there any golang library that can be used for git commands execution? or exec is the only way. Also @StrongMonkey pointed that we have this util that can be used: https://github.com/rancher/rancher/blob/a4499e242f333661453a7e0cb7a6e926a1526c32/pkg/catalog/git/git.go#L34

Reply:
I think the most mature one would be `https://github.com/src-d/go-git`. Though it's not a feature complete substitution for git command yet, it should be feasible to cover the use cases here, so I will try with that.

Reply:
The `go-git` library has no equivalent of a `git ls-remote url branch` so it will need to list all remotes of a repo and may take twice or triple time comparing with exec. I would like to update `pkg/catalog/git/git.go` until we find some better solution.